modules/graph/vertex_map/arrow_vertex_map.h

```diff
@@ -49,20 +50,38 @@ class ArrowVertexMap
                 "Expect arrow_string_view in vertex map's OID_T");
 
  public:
-  ArrowVertexMap() {}
+  explicit ArrowVertexMap(bool use_perfect_hash)
+      : use_perfect_hash_(use_perfect_hash) {}
   ~ArrowVertexMap() {}
 
   static std::unique_ptr<vineyard::Object> Create() __attribute__((used)) {
     return std::static_pointer_cast<vineyard::Object>(
         std::unique_ptr<ArrowVertexMap<OID_T, VID_T>>{
-            new ArrowVertexMap<OID_T, VID_T>()});
+            new ArrowVertexMap<OID_T, VID_T>(false)});
   }
 
   void Construct(const vineyard::ObjectMeta& meta);
 
   bool GetOid(vid_t gid, oid_t& oid) const;
 
-  bool GetGid(fid_t fid, label_id_t label_id, oid_t oid, vid_t& gid) const;
+  bool GetGid(fid_t fid, label_id_t label_id, oid_t oid, vid_t& gid) const {
+    if (use_perfect_hash_) {
+      const std::pair<OID_T, VID_T>* res = o2g_p_[fid][label_id].find(oid);
+      if (res) {
+        gid = res->second;
+        delete res;
+        return true;
+      }
+      return false;
+    } else {
+      auto iter = o2g_[fid][label_id].find(oid);
+      if (iter != o2g_[fid][label_id].end()) {
+        gid = iter->second;
+        return true;
+      }
+      return false;
+    }
+  }
```

Comment:
Move to `impl.h` to optimize compile time.

Reply:
Done